### PR TITLE
Name the refusal where the failure was anonymous

### PR DESCRIPTION
**Pull request by jmrplens**
An audit recorded the call sites where a bad input dies deep inside stdlib or numpy, in words that name neither the parameter nor the function: a `KeyError` spelling out a dict key, an `IndexError` about axis 0, LAPACK announcing that an SVD did not converge because a NaN reached it. These are observable failures of a published library, so changing them was held for an explicit decision; with that approval, 93 of the surviving sites now refuse by name at the entry point the caller actually typed, and the other six were measured and found already fixed by earlier waves.

Reproducing each failure before touching it is what the batch was for, and it paid in the usual currency: several sites were not loud and wrong but silent and wrong. A misspelt `label_side="rigth"` rendered byte-identically to the `left` it fell back to, in a module whose stated point is that a typo must break generation readably. A temperature below absolute zero returned an absorption area of minus 127 square metres behind a warning. An infinite airflow input returned an all-NaN result without a word, and a four-entry room dimension was silently truncated to three.

The guards use the shared validation helpers where one fits and the house voice where none does, and they sit at the public entry point rather than in the private helper the traceback used to end in. Two judgement calls worth naming: a scalar `frequencies` is accepted as the single band's centre and refused by name for several bands, the same reading the integer-order precedent takes of a value that arrived through arithmetic; and one site avoided `require_positive_array` because its coercion would have changed which scalars the function accepts, which is more than the finding licensed.

Every guard carries a refusal test matched on the parameter name, and the tests that pinned the old anonymous failures now pin the named ones. Fourteen reference pages move because the docstrings gained the `:raises` lines the guards now honour.

<a href="https://cubic.dev/pr/jmrplens/phonometry/pull/631?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened validation for invalid, non-finite, mismatched, and malformed inputs across acoustics, filters, materials, propagation, simulation, and psychoacoustics.
  * Replaced low-level errors and misleading results with clearer parameter-specific messages.
  * Improved array broadcasting, frequency-range checks, plotting-axis requirements, and handling of short signals and list-based attenuation data.

* **Documentation**
  * Expanded API documentation for accepted inputs and `ValueError` conditions.

* **Tests**
  * Added broad regression coverage for validation, edge cases, broadcasting, plotting, and short-signal processing.

### CODE REVIEW

File: src/phonometry/building/measurement/uncertainty.py [hunk 1]
```diff
@@ -713,9 +713,12 @@ def prediction_input_uncertainty(
...
     :param sigma_product: Product-homogeneity standard deviation ``σ_product``, in dB.
     :param n: Number of measurements of the product (:math:`n \ge 1`).
-    :raises ValueError: Non-positive ``n`` or a negative standard deviation.
+    :raises ValueError: A non-integer or non-positive ``n``, or a negative
+        standard deviation.
     """
-    if n < 1:
+    # A fractional or NaN ``n`` sails past a bare ``n < 1`` and is divided
+    # by, so integrality is required explicitly (numpy integers included).
+    if not isinstance(n, (int, np.integer)) or n < 1:
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/building/measurement/uncertainty.py:721]: I reproduced this and measured it: prediction_input_uncertainty(1.2, 1.0, True) returns 1.8547236990991407, bit-identical to prediction_input_uncertainty(1.2, 1.0, 1). In Python True == 1 by the language, so the boolean behaves as exactly one measurement and the formula prints the correct n=1 answer, not a silently wrong one. My policy across this review is to reject bool-policing unless the boolean produces a wrong result: a caller writing n=True is not plausible arithmetic the way n=3.0 is, and that case (a fractional n) is already refused by name, which I verified: n=1.0 raises ValueError 'n must be a positive integer.' Not adding an isinstance(bool) check here.

File: src/phonometry/electroacoustics/frequency_response.py [hunk 2]
```diff
@@ -100,6 +101,35 @@ def _spectra(
...
+        not isinstance(nperseg, (int, float, np.integer, np.floating))
+        or not math.isfinite(nperseg)
+        or int(nperseg) != nperseg
+    ):
+        msg = "'nperseg' must be a positive integer."
+        raise ValueError(msg)
+    seg = _default_nperseg(n, fs) if nperseg is None else int(nperseg)
+    if seg < _MIN_SAMPLES or seg > n:
+        msg = "'nperseg' must be between 32 and the signal length."
+        raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/electroacoustics/frequency_response.py:133]: Reproduced from the public entry point: transfer_function(x, x, 48000.0, overlap=10**10000) and transfer_function(x, x, 48000.0, nperseg=10**10000) both printed OverflowError: int too large to convert to float, escaping the guard's own named ValueError contract. Fixed in _welch_params, not by catching OverflowError but by keeping the checks total: the overlap comparison now runs in exact int/float arithmetic (0.0 <= overlap < 1.0 without the float() call, which Python evaluates exactly for arbitrary-size ints), and math.isfinite is probed only for floats, since integers are finite by construction. After the fix I measured: overlap=10**10000 prints ValueError: 'overlap' must be in [0, 1). and nperseg=10**10000 prints ValueError: 'nperseg' must be between 32 and the signal length. Both refusals name the parameter. Added two regression tests beside the existing refusal tests; ruff, mypy (455 files clean) and the module suite (167 passed) are green.

File: src/phonometry/emission/sound_power_intensity.py [hunk 3]
```diff
@@ -1178,6 +1182,27 @@ def precision_field_indicators(
...
+    """The band centres validated against the indicator's band axis.
+
+    Split out of :func:`precision_qualification`, which was one branch past
+    what one reader holds: the centre validation is one idea, and it reads
+    the same whether the caller supplied centres or not.
+    """
+    if frequencies is None:
+        return None
+    freqs = require_positive_array(frequencies, "frequencies")
+    if freqs.shape != f_pi_signed.shape:
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/emission/sound_power_intensity.py:1197]: I measured what actually escapes: require_positive_array(("bad",), "frequencies") raises ValueError: could not convert string to float: 'bad', and require_per_band prints the same for a string tuple; an object() input raises TypeError: float() argument must be a string or a real number, not 'object'. So the input is refused loudly, the exception class honors the documented ValueError contract for string-like garbage, numpy's message names the offending value, and TypeError is the right class for the genuinely wrong-typed case. Wrapping np.asarray in both shared validators would duplicate a refusal numpy already makes by name, and the fix site is src/phonometry/_internal/validation.py, a helper shared by dozens of modules across the tree, so renaming its conversion errors is a deliberate cross-tree pass, not a per-entry-point guard extension. The mistakes a real caller plausibly makes with frequencies or scan intensities (negative, zero, NaN, wrong length) all get parameter-named refusals already, which I verified are untouched. Rejecting.

File: src/phonometry/filters/weighting.py [hunk 4]
```diff
@@ -639,7 +659,12 @@ def _prepare_time_weighting_initial_state(
             return np.asarray(np.take(x_sq, 0, axis=-1), dtype=x_sq.dtype).copy()
         raise ValueError(invalid_initial_state_message)
 
-    state = np.asarray(initial_state, dtype=x_sq.dtype)
+    try:
+        state = np.asarray(initial_state, dtype=x_sq.dtype)
+    except TypeError:
+        # numpy's float() message would name neither the parameter nor the
+        # accepted forms; the module's own message does both.
+        raise TypeError(invalid_initial_state_message) from None
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/filters/weighting.py:668]: Reproduced. I ran time_weighting(x, 48000, mode="fast", initial_state=["invalid"]) on numpy 2.5.2 and it printed numpy's anonymous ValueError: could not convert string to float: 'invalid', while initial_state=object() already got the module's named TypeError. The guard existed precisely to rename numpy's conversion failures, so I widened the except to (TypeError, ValueError); both now raise the module TypeError naming the parameter and the accepted forms, and I added a refusal test beside the dict one. On bools: initial_state=True converts to a valid scalar state of 1.0, which is ordinary arithmetic, so it is not policed.

File: src/phonometry/filters/weighting.py [hunk 5]
```diff
@@ -854,13 +881,27 @@ def linkwitz_riley(
...
+    freq = require_positive(freq, "freq")
+    nyquist = fs / 2
+    if freq >= nyquist:
+        msg = f"'freq' must be below the Nyquist frequency ({nyquist:g} Hz); got {freq:g}."
+        raise ValueError(msg)
 
     # A Linkwitz-Riley filter of order N is two Butterworth filters of order N/2 in series
     half_order = order // 2
-    wn = freq / (fs / 2)
+    wn = freq / nyquist
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/filters/weighting.py:905]: This does not reproduce. I ran linkwitz_riley(x, 48000, freq=1000.0, order=2.0) on scipy 1.18.0 and it returned normally; signal.butter(1.0, 0.1, btype="low", output="sos") also succeeds, so no dependency error exists to pre-empt. I then compared order=2.0/4.0/6.0 against their int counterparts and np.array_equal printed True for every band, so the float order produces bit-identical, correct output. Every non-integral float (2.5, 3.0, 4.5) is already refused by name by the parity check (x % 2 != 0 for any non-integral value), and on bools order=True is refused by parity and order=False by the positivity check, both with the named message, so no bool policing is needed either. There is no input here that reaches SciPy un-named or produces a wrong result, so I left the guard as it is.

File: src/phonometry/materials/absorbers/sound_absorption.py [hunk 6]
```diff
@@ -145,12 +149,19 @@ def _resolve_speed(temperature: float, speed_of_sound: float | None) -> float:
...
     Warns when the speed is derived from a temperature outside the 15..30 degC
     validity range of Eq. (6). An explicit ``speed_of_sound`` bypasses the check.
+    Rejects a physically impossible ``temperature`` outright: Eq. (6) applied
+    below absolute zero would hand back a zero or negative speed of sound.
     """
     if speed_of_sound is not None:
         if speed_of_sound <= 0.0:
             msg = "'speed_of_sound' must be positive."
             raise ValueError(msg)
         return float(speed_of_sound)
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/materials/absorbers/sound_absorption.py:161]: Reproduced, and both failures were silent. I ran absorption_coefficient(t1, t2, 200.0, 10.0, speed_of_sound1=math.nan) and it printed [nan nan]; with math.inf it printed [0. 0.], because infinity is positive so it passed the <= 0 check and zeroed the speed-dependent term of Eq. (5). measure_sound_absorption with speed_of_sound=nan likewise returned NaN alpha_s. The neighbouring temperature guard already names NaN, so I extended this one to match: not math.isfinite(speed_of_sound) or speed_of_sound <= 0.0, with the message "'speed_of_sound' must be finite and positive." (the old "must be positive" would have misnamed the refusal of inf). Added refusal tests for nan, inf and a negative speed beside the temperature ones. On bools: speed_of_sound=True is the finite positive number 1.0 and passes as plain arithmetic, so it is not policed.

File: src/phonometry/materials/absorbers/sound_absorption.py [hunk 7]
```diff
@@ -636,6 +648,18 @@ def measure_sound_absorption(
...
+    else:
+        # A non-numeric humidity becomes NaN here so that it fails the range
+        # test below and is refused by name, instead of dying inside float().
+        try:
+            humidity_pct = float(humidity)
+        except ValueError:
+            humidity_pct = math.nan
+        if not 0.0 <= humidity_pct <= _MAX_RELATIVE_HUMIDITY_PERCENT:
+            msg = "'humidity' must be within [0, 100] %."
+            raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/materials/absorbers/sound_absorption.py:665]: Reproduced. I ran measure_sound_absorption with humidity=np.array([54.0]) and it printed TypeError: only 0-dimensional arrays can be converted to Python scalars; humidity=[] and humidity=object() printed float()'s TypeError too, all bypassing the named ValueError the comment above the guard promises. A one-element array sliced out of logged climate data is exactly the kind of input a real caller produces, so I widened the except to (TypeError, ValueError); those inputs now become NaN, fail the range test, and are refused with "'humidity' must be within [0, 100] %." like the string case. Added a refusal test with np.array([54.0]) beside the "wet" one. On bools: humidity=True is 1.0 %, a value inside the range that gets recorded as ordinary arithmetic, so it is not policed.

File: src/phonometry/aircraft/certification.py [hunk 8]
```diff
@@ -379,12 +379,21 @@ def tone_correction(
...
+    # anything the index cannot be (2.9, "aeroplane", None) is refused by name
+    # rather than left to fail inside int().
+    try:
+        start = int(start_band)
+        integral = start == start_band
+    except (TypeError, ValueError, OverflowError):
+        start, integral = -1, False
+    if not integral or not 0 <= start <= _MAX_TONE_START_BAND:
+        msg = "'start_band' must be an integer between 0 (50 Hz) and 3 (100 Hz)."
         raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/aircraft/certification.py:395]: I ran this against the pinned NumPy 2.5.2 in the project venv: `tone_correction(spl, start_band=np.array([2]))` raises `ValueError: 'start_band' must be an integer between 0 (50 Hz) and 3 (100 Hz).` It cannot slip through, because on 2.5.2 `int(np.array([2]))` raises `TypeError: only 0-dimensional arrays can be converted to Python scalars` (the ndim > 0 conversion deprecated in 1.25 is now an error), and the guard's `except (TypeError, ValueError, OverflowError)` turns that into the named refusal. I also measured `np.array([2.5])` and `np.array([2, 3])`: both refused with the same message. The project floor is `numpy>=2.5.2` in pyproject.toml and requirements.txt pins 2.5.2, so no supported NumPy reaches the accepting path your sandbox saw; that run must have used an older NumPy where the conversion still succeeded. A 0-d `np.array(2)` does convert and returns the same value as `start_band=2`, which is the documented integral-value rule at work, same as `start_band=2.0`. On the same rule, `start_band=True` converts to index 1 and computes the correct band, so I am not policing booleans here either. No change made.

File: src/phonometry/building/measurement/structure_borne_power.py [hunk 9]
```diff
@@ -192,6 +194,14 @@ def structure_borne_power_level(
...
         msg = "'loss_factor' must contain positive, finite values."
         raise ValueError(msg)
+    try:
+        np.broadcast_shapes(lv.shape, f.shape, eta.shape)
+    except ValueError:
+        msg = (
+            "'velocity_level', 'frequency' and 'loss_factor' must broadcast "
+            f"together; got shapes {lv.shape}, {f.shape} and {eta.shape}."
+        )
+        raise ValueError(msg) from None
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/building/measurement/structure_borne_power.py:209]: Reproduced: `structure_borne_power_level(70.0, np.nan, 100.0, 2.0, 0.05)` returned `nan` and `np.inf` returned `inf`, because NaN compares False against `f <= 0.0`, and `velocity_level=np.nan` sailed through with no check at all and printed `nan`. That is a silently wrong result from a measured quantity, and this file already refuses non-finite levels by name in `source_mobility_from_levels`, so I extended the guard in the same voice: `frequency` now raises `'frequency' must contain positive, finite values.` (matching the neighbouring `loss_factor` wording) and `velocity_level` raises `'velocity_level' must contain only finite values.` before the broadcast check. Both refusals verified for nan and inf; the happy path still returns 54.9715 dB for the 500 Hz hand case. Refusal tests added beside the existing ones; the module's suite passes 62, mypy is clean over src and scripts, ruff is clean.

File: src/phonometry/building/measurement/structure_borne_power.py [hunk 10]
```diff
@@ -408,6 +419,14 @@ def equivalent_blocked_force_level(
...
         msg = "'plate_mobility' must be finite with a positive real part."
         raise ValueError(msg)
+    try:
+        np.broadcast_shapes(lw.shape, y.shape)
+    except ValueError:
+        msg = (
+            "'power_level' and 'plate_mobility' must broadcast together; "
+            f"got shapes {lw.shape} and {y.shape}."
+        )
+        raise ValueError(msg) from None
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/building/measurement/structure_borne_power.py:438]: Reproduced at both entry points: `equivalent_blocked_force_level(np.nan, 1e-6)` and `equivalent_free_velocity_level(np.nan, 1e-4+1e-5j)` returned `nan`, and inf returned `inf`, while `plate_mobility` next to them was already checked finite by name. I added the missing half of the pair at both Formula (15) and Formula (18): `power_level` is now validated right after coercion and raises `'power_level' must contain only finite values.`, the exact wording `source_mobility_from_levels` already uses for its level inputs, with the docstrings' raises clauses updated. Both refusals verified for nan and inf at both sites. A refusal test covering the four cases sits beside the existing shape-mismatch tests; the module's suite passes 62, the downstream EN 12354-5 prediction tests pass 32, mypy is clean over src and scripts, ruff is clean.

File: src/phonometry/noise_control/hvac.py [hunk 11]
```diff
@@ -827,7 +838,7 @@ def blade_passing_frequency(rotational_speed: float, blades: int) -> float:
     :raises ValueError: If ``blades`` is not a positive integer.
     """
     rpm = require_positive(rotational_speed, "rotational_speed")
-    if blades <= 0:
+    if blades <= 0 or not float(blades).is_integer():
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/noise_control/hvac.py:841]: I reproduced this and measured what a bool actually does at all five sites. blade_passing_frequency(1200, True) printed 20.0, identical to blades=1. silencer_self_noise(None, 10.0, True, 0.6) and diffuser_sound_power(None, 0.1, 0.05, 20.0, count=True) both compared array-equal to the passages=1 and count=1 results. audible_image_count(True) printed 6, same as order 1, and image_source_rir(..., max_order=True) produced an RIR array-equal to max_order=1 with the stored max_order normalised to int 1. In Python True == 1, and every one of these results is exactly the correct answer for one blade, one passage, one device, order one; nothing goes silently wrong. My rule for this batch is to reject bool-policing unless the boolean silently produces a wrong result, and a caller writing n=True is not plausible arithmetic the way n=3.0 is. The one blemish is cosmetic: the silencer label prints N = True instead of N = 1, which changes no number. Rejecting as input-policing for a caller that does not exist.

File: src/phonometry/psychoacoustics/quality/roughness_ecma.py [hunk 12]
```diff
@@ -635,7 +635,7 @@ def roughness_ecma(
         msg = "fs must be positive"
         raise ValueError(msg)
     if fs != _FS:
-        x = signal.resample(x, round(x.size * _FS / fs))
+        x = signal.resample(x, max(1, round(x.size * _FS / fs)))
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/psychoacoustics/quality/roughness_ecma.py:641]: Reproduced, and it was worse than a stray NaN in the output: 1 s of calibrated noise with x[100] = NaN returned roughness 0.0 with an all-finite specific-roughness array, and an inf sample likewise printed 0.0. The poison washes out in the envelope stages and the result is indistinguishable from silence, which is exactly the silent-garbage case a guard is for. The loudness entry point of the same standard already refuses this by name (loudness/ecma.py, "'signal_in' must be finite."), so roughness_ecma was the odd one out. I added the identical guard after the empty-input check and before resampling, plus test_non_finite_signal_raises covering NaN and inf beside the empty-signal refusal. ruff clean, mypy "Success: no issues found in 455 source files", and the module suites print 204 passed.

File: src/phonometry/room/acoustics.py [hunk 13]
```diff
@@ -588,6 +593,9 @@ def decay_curve(
         if band <= 0.0:
             msg = "Band centre frequency 'band' must be positive."
             raise ValueError(msg)
+        if fraction <= 0:
+            msg = "Bandwidth 'fraction' must be positive."
+            raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/room/acoustics.py]: I ran all three non-finite fractions through decay_curve on a synthetic exponential IR. fraction=inf reproduced your claim exactly: it raised "The lower limit must be less than the upper limit.", naming the bank's limits instead of fraction (the half-width collapses to 1). fraction=nan was already refused by name downstream ("'fraction' must be positive." out of require_positive in the bank's frequency generation), and -inf was caught by the existing entry guard. Only +inf escaped with a misnamed refusal, so I applied the one-clause extension: not np.isfinite(fraction) or fraction <= 0, keeping the existing message, which matches how the shared require_positive helper words the non-finite case everywhere else. Added test_non_finite_fraction_is_refused for inf and nan beside the non-positive refusal. ruff clean, mypy "Success: no issues found in 455 source files", and the suites print 204 passed.

File: src/phonometry/underwater/sources/pile_driving_noise.py [hunk 14]
```diff
@@ -261,10 +261,19 @@ def strike_sel_spectrum(
     fs = resolve_fs(pressure, fs, name="pressure")
     sig = _validate_pressure(pressure, min_samples=2)
     fs_v = _positive(fs, "fs")
-    if int(fraction) not in (1, 3):
+    # The value the caller passed, not its truncation: int() would silently
+    # turn 3.9 into one-third octaves and 1.4 into octaves.
+    if fraction not in (1, 3):
         msg = "'fraction' must be 1 (octave) or 3 (one-third octave)."
         raise ValueError(msg)
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/underwater/sources/pile_driving_noise.py:268]: I reproduced the claim before deciding: strike_sel_spectrum(p, 48000.0, fraction=np.array([1, 3])) raises numpy's "The truth value of an array with more than one element is ambiguous" ValueError, and a one-element np.array([3]) passes the membership test only to die later in int(fraction). I am not adding the scalar check. 'fraction' is a scalar mode selector (1 or 3, default 3); no real caller passes an array of band fractions, and the two sibling guards in the tree (emission/intensity.py _validate_band_options and filters/frequencies.py) make the same plain membership test, so policing arrays in this one spot would guard an input nobody produces and break the convention. Numpy integer scalars and 0-d arrays still pass and compute correctly, which is the array-adjacent input a caller can actually produce. On booleans: fraction=True is accepted as 1 (octave bands) because True == 1 in Python; that yields the documented octave result, not a silently wrong number, so I am not policing bools here either.

File: src/phonometry/underwater/sources/ship_radiated_noise.py [hunk 15]
```diff
@@ -94,14 +96,12 @@ def hydrophone_depths(
...
-        ang.size == 0
-        or not np.all(np.isfinite(ang))
-        or np.any(ang <= 0.0)
-        or np.any(ang >= _VERTICAL_DEG)
-    ):
-        msg = "'angles' must be finite and lie in the open interval (0, 90) degrees."
+    # The shared guard also pins the shape: a nested pair of pairs used to
+    # come back as a 2-D "depths" array, and a bare scalar as a 0-d one whose
+    # documented per-hydrophone read depths[0] died in numpy.
+    ang = require_positive_array(angles, "angles")
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/underwater/sources/ship_radiated_noise.py:102]: Reproduced: hydrophone_depths(100.0, ("bad",)) escaped as numpy's anonymous "could not convert string to float: 'bad'", and a dict escaped as a bare TypeError, neither naming the parameter. I wrapped the float64 conversion in both shared validators (require_positive_array and require_finite_array in phonometry/_internal/validation.py) so a conversion failure now raises the parameter-named refusal, in the same voice the tree already uses for fresnel_number. After the fix I ran the same calls and they print: ValueError: 'angles' must be numeric. / 'frequency' must be numeric. / 'rnl' must be numeric. (the dict case now also names 'angles'). Refusal tests sit beside the existing ones in test_ship_radiated_noise.py, covering angles through hydrophone_depths and frequency and rnl through monopole_source_level. Gates: ruff clean, mypy prints Success: no issues found in 455 source files, and since the validators are shared I ran the suites of every consuming module: 4944 passed.

File: tests/building/measurement/test_structure_borne_power.py [hunk 16]
```diff
@@ -339,3 +339,56 @@ def test_power_level_rejects_nonpositive_loss_factor() -> None:
...
+def test_power_level_rejects_mismatched_band_axes() -> None:
+    """A velocity level of the wrong band count is refused by name."""
+    with pytest.raises(
+        ValueError,
+        match="'velocity_level', 'frequency' and 'loss_factor' must broadcast",
+    ):
+        building.structure_borne_power_level(
+            [80.0, 82.0], [500.0, 1000.0, 2000.0], 10.0, 1.0, 0.01
+        )
+
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply tests/building/measurement/test_structure_borne_power.py:397]: Reproduced against the reviewed revision: structure_borne_power_level(80.0, float("nan"), 10.0, 1.0, 0.01) returned nan and float("inf") returned inf, silently, exactly as claimed, because NaN passes a <= 0 comparison and infinity counts as positive. The frequency guard now tests finiteness as well as sign, in the same voice as the loss_factor guard beside it: both calls now print ValueError: 'frequency' must contain positive, finite values. Regression cases for NaN and +inf (and a NaN velocity level) live in test_power_level_rejects_non_finite_frequency_and_velocity beside the existing refusal tests. The module's tests pass: the run over the batch's four test files prints 96 passed in 1.51s.

File: src/phonometry/emission/sound_power_intensity.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/emission/sound_power_intensity.py:1197]: Retracting my rejection above: a parallel batch answering the neighbouring threads implemented exactly this, a shared `_as_float64` that names the parameter when numpy cannot convert, and it had landed in the two array validators between that measurement and now. With the idea in the tree the consistency argument flips, so I have completed it: all four coercing validators (`require_positive_array`, `require_finite_array`, `require_per_band`, `require_1d_signal`) now route their conversion through it, and a parametrized test drives tuple-of-strings, object and dict garbage through each, asserting the named refusal. Measured after: every one prints `'<name>' must be numeric.` with numpy's original error chained underneath.
